### ISSUE TITLE
Postgres implementation to include WINDOW functions?

### ISSUE THREAD

**Issue by nukoyluoglu** [sqlancer/sqlancer#76]
I believe the SELECT queries generated for testing by the Postgres implementation do not involve WINDOW clauses right now. I was wondering if they will be incorporated into the implementation, as well as other clauses that are currently not supported.

**Comment by mrigger** [comment]
Yes, you are right, SQLancer currently does not generate window functions for Postgres. They are generated for some of the other DBMS, like SQLite3, but SQLancer does currently also not provide a test oracle to determine whether the window functions' result is correct, limitting the usefulness of generating them.

I do not plan to implement window functions and other features for Postgres in the near future. Contributions in this direction are very welcome, though!

**Comment by EthanDBer** [comment]
I want to try this. If possible, please assign me. 
BTW, 
Currently in SQLancer, which database is more comprehensive in terms of window function support? 
Please give me some hints, I believe I can solve this problem.

**Comment by cv-prasad** [comment]
@mrigger  can you assign me this issue. I would like to work on this.

**Comment by chaitanyashetty47** [comment]
@mrigger can I please look into this?

**Comment by mrigger** [comment]
Not sure if it's useful to do so. If you are looking to contribute to the project, one way would be to update an existing implementation to the latest version of the DBMS. For example, the MariaDB implementation is outdated.

**Comment by chaitanyashetty47** [comment]
Surely :) I am actually new to your project and also to open source. Any suggestions on where should I start with to get a basic idea about SQLancer, so that I could work on the MariaDB implementation?

**Comment by Avi1832** [comment]
If you want to add WINDOW functions support in the PostgreSQL implementation of SQLancer, you will need to change the query generation mechanism to incorporate window functions in SQL test cases. Here's one way to add them:

Steps to Add WINDOW Functions to SQLancer (PostgreSQL Implementation)

#1. Alter Query Generation to Add Window Functions
You should change the SQL generation mechanism in SQLancer's PostgreSQL implementation to incorporate proper WINDOW functions.

Sample Functions to Add:

ROW_NUMBER() OVER (PARTITION BY col ORDER BY col)
RANK() OVER (ORDER BY col)
DENSE_RANK() OVER (PARTITION BY col ORDER BY col DESC)
SUM(col) OVER (PARTITION BY col)
AVG(col) OVER (ORDER BY col ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)
These need to be included in:

PostgresExpressionGenerator.java (or the file responsible for query generation)
The logic needs to ensure random valid window function generation.

#2. Ensure Window Functions Are Used in Queries
After window functions are created, they must be applied in SELECT statements. Example:

sql
Copy
Edit
SELECT id, salary,
       RANK() OVER (ORDER BY salary DESC) AS salary_rank
FROM employees;
You can change the test case generator to generate random but correct window queries.


#3. Process Expected Results & Bugs
SQLancer typically identifies bugs by verifying whether two queries provide inconsistent results.
When testing window functions, make sure query results are deterministic over multiple executions.
Possible SQLancer Checks:
Compare window function outputs with corresponding GROUP BY queries (if applicable).
Make sure there are no crashes or abnormal behavior.

#4. Update Expected Behavior Validation
SQLancer must ensure that PostgreSQL doesn't crash and results are deterministic.
You might need to modify ExpectedErrors.java to add known edge cases for window functions.#1081 

#5. Test with Various PostgreSQL Versions
Support various PostgreSQL versions (at least PostgreSQL 11+, where WINDOW functions are natively supported).

**Comment by Eklavya-Sahani** [comment]
@mrigger If you're curious about the technical aspects, here's a general idea of how WINDOW clause generation could be added to the Postgres provider in SQLancer:

Extend the Query Generator: The core of the Postgres provider has classes responsible for generating different parts of a SELECT statement (e.g., columns, FROM clause, WHERE clause, GROUP BY, ORDER BY). New code would need to be added to generate WINDOW clauses. This would involve:
Randomly deciding whether to include a WINDOW clause.
Generating unique window names.
Randomly deciding on the window specification:
PARTITION BY clause (including generating expressions for partitioning).
ORDER BY clause within the window (including generating expressions and ASC/DESC).
ROWS or RANGE framing clauses (including generating bounds like UNBOUNDED PRECEDING, CURRENT ROW, n PRECEDING, n FOLLOWING, etc.).
Optionally referencing other named windows.
Integrate with Column Generation: When generating the actual SELECT list, the generator would need to be able to randomly choose to include window functions (e.g., ROW_NUMBER(), RANK(), LAG(), LEAD(), aggregate functions with OVER(...)). The OVER(...) clause would then reference one of the generated WINDOW clauses.
Consider Semantic Validity: While SQLancer aims to generate a wide variety of queries, some basic semantic rules need to be followed to avoid trivially invalid queries that the database would immediately reject. For WINDOW clauses, this might involve ensuring that the expressions used in PARTITION BY and ORDER BY are valid for the involved tables and columns.
Add Configuration Options (Optional): To give users more control, options could be added to the command-line interface to influence the probability of generating WINDOW clauses or to enable/disable specific aspects of window function generation.

**Comment by robins** [comment]
@Eklavya-Sahani -  If you're interested in implementing the feature, well-rounded PRs are always welcome.


However, if you're starting on the journey, (as mentioned above), it may be simpler to improvise one of the other engines (where part of this feature is already implemented)
